actix-session/src/storage/sqlite.rs

```diff
@@ -0,0 +1,348 @@
+    utils::generate_session_key,
+    SessionKey, SessionStore,
+};
+
+/// Use Sqlite as session storage backend.
+///
+/// ```no_run
+/// use actix_session::{storage::SqliteSessionStore, SessionMiddleware};
+/// use actix_web::{
+///   cookie::Key,
+///   web, App, HttpServer, HttpResponse
+/// };
+/// use r2d2_sqlite::{self, SqliteConnectionManager};
+///
+/// // The secret key would usually be read from a configuration file/environment variables.
+/// fn get_secret_key() -> Key {
+///     # todo!()
+///     // [...]
+/// }
+///
+/// #[actix_web::main]
+/// async fn main() -> std::io::Result<()> {
+///   let secret_key = get_secret_key();
+///
+///   let manager = SqliteConnectionManager::file("sessions.db");
+///   let pool = r2d2::Pool::<r2d2_sqlite::SqliteConnectionManager>::new(manager).unwrap();
+///   let store = SqliteSessionStore::new(pool, true).unwrap();
+///
+///   HttpServer::new(move || {
+///       App::new()
+///           .wrap(SessionMiddleware::new(
+///                 store.clone(),
+///                 secret_key.clone()
+///            ))
+///           .default_service(web::to(|| HttpResponse::Ok()))
+///   })
+///   .bind(("127.0.0.1", 8080))?
+///   .run()
+///   .await
+/// }
+/// ```
+///
+/// # Implementation notes
+/// `SqliteSessionStore` uses rusqlite, r2d2 and r2d2_sqlite.
+///
+/// [`rusqlite`]: https://github.com/rusqlite/rusqlite
+/// [`r2d2`]: https://github.com/sfackler/r2d2
+/// [`r2d2_sqlite`]: https://github.com/ivanceras/r2d2-sqlite
+#[cfg_attr(docsrs, doc(cfg(feature = "sqlite-session")))]
+#[derive(Clone)]
+pub struct SqliteSessionStore {
+    configuration: CacheConfiguration,
+    pool: Pool<SqliteConnectionManager>,
+}
+
+#[derive(Clone)]
+struct CacheConfiguration {
+    cache_keygen: Arc<dyn Fn(&str) -> String + Send + Sync>,
+}
+
+impl Default for CacheConfiguration {
+    fn default() -> Self {
+        Self {
+            cache_keygen: Arc::new(str::to_owned),
+        }
+    }
+}
+
+impl SqliteSessionStore {
+    /// Create a new instance of [`SqliteSessionStore`] using the default configuration.
+    /// It takes two required inputs to create a new instance of [`SqliteSessionStore`] - a
+    /// pool of connections [`Pool<SqliteConnectionManager>`] and a boolean which specifies
+    /// whether default garbage connection is enabled/disabled.
+    ///
+    /// Default garbage collection of stale sessions happens through a database trigger, which
+    /// executes before every insert on the `sessions` table.
+    pub fn new(
+        pool: Pool<SqliteConnectionManager>,
+        garbage_collect: bool,
+    ) -> Result<Self, anyhow::Error> {
+        // create sessions table if it doesn't exist
+        let conn = pool.get().unwrap();
+        conn.execute(
+            "CREATE TABLE IF NOT EXISTS sessions (
+          id INTEGER NOT NULL,
+          session_key	TEXT NOT NULL,
+          session_state	TEXT NOT NULL,
+          expiry INTEGER NOT NULL,
+          PRIMARY KEY(id AUTOINCREMENT)
```

Comment:
This allows for multiple rows sharing the same `session_key` - that's not what we want. I'd suggest using the `session_key` as primary key or, at the very least, add a `UNIQUE` constraint on it.